`set -o pipefail` is part of POSIX now, consider disabling the warning for `sh`

The POSIX:2022 specification adds support for `set -o pipefail`, so it might make sense to disable this warning even for `#/bin/sh`. 

Source on adding pipefail: POSIX discussion <https://www.austingroupbugs.net/view.php?id=789>, list of shell support <https://unix.stackexchange.com/questions/654885/who-is-responsible-for-providing-set-o-pipefail/654932#654932>, work on `dash` adding it <https://lore.kernel.org/dash/20220422211013.18195-1-chris@chrisn.me.uk/>

As mentioned in stackoverflow, most `sh` implementations support pipefail already, `dash` being about the last big one to not support it. And they are working on it, as mentioned above.

#### For bugs
- Rule Id (if any, e.g. SC1000):  SC3040 (pipefail)
- My shellcheck version (`shellcheck --version` or "online"):  v0.8.0
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit


#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
set -o pipefail
```

#### Here's what shellcheck currently says:

```
set -o pipefail
       ^-- [SC3040](https://www.shellcheck.net/wiki/SC3040) (warning): In POSIX sh, set option pipefail is undefined.
```

#### Here's what I wanted or expected to see:

No error

FYI, another workaround for now is to just try setting pipefail in a subshell (so failures don't propagate), and then set it in your main shell only if it succeeds. I added this to the wiki

```sh
# shellcheck disable=SC3040
(set -o pipefail 2> /dev/null) && set -o pipefail
```

Austin Group released draft 4.1 last month, hopefully they are getting close to final... https://www.opengroup.org/austin/

--- a/src/ShellCheck/AnalyzerLib.hs
+++ b/src/ShellCheck/AnalyzerLib.hs
@@ -224,9 +224,9 @@ makeParameters spec = params
         hasPipefail =
             case shellType params of
                 Bash -> isOptionSet "pipefail" root
-                Dash -> True
+                Dash -> isOptionSet "pipefail" root
                 BusyboxSh -> isOptionSet "pipefail" root
-                Sh   -> True
+                Sh -> isOptionSet "pipefail" root
                 Ksh  -> isOptionSet "pipefail" root,
         hasExecfail =
             case shellType params of

--- a/src/ShellCheck/Checks/ShellSupport.hs
+++ b/src/ShellCheck/Checks/ShellSupport.hs
@@ -175,7 +175,7 @@ prop_checkBashisms80 = verifyNot checkBashisms "#!/bin/sh\nhash -r"
 prop_checkBashisms81 = verifyNot checkBashisms "#!/bin/dash\nhash -v"
 prop_checkBashisms82 = verifyNot checkBashisms "#!/bin/sh\nset -v +o allexport -o errexit -C"
 prop_checkBashisms83 = verifyNot checkBashisms "#!/bin/sh\nset --"
-prop_checkBashisms84 = verify checkBashisms "#!/bin/sh\nset -o pipefail"
+prop_checkBashisms84 = verifyNot checkBashisms "#!/bin/sh\nset -o pipefail"
 prop_checkBashisms85 = verify checkBashisms "#!/bin/sh\nset -B"
 prop_checkBashisms86 = verifyNot checkBashisms "#!/bin/dash\nset -o emacs"
 prop_checkBashisms87 = verify checkBashisms "#!/bin/sh\nset -o emacs"
@@ -383,8 +383,8 @@ checkBashisms = ForShell [Sh, Dash, BusyboxSh] $ \t -> do
         beginsWithDoubleDash = (`matches` mkRegex "^--.+$")
         longOptions          = Set.fromList
             [ "allexport", "errexit", "ignoreeof", "monitor", "noclobber"
-            , "noexec", "noglob", "nolog", "notify" , "nounset", "verbose"
-            , "vi", "xtrace" ]
+            , "noexec", "noglob", "nolog", "notify" , "nounset", "pipefail"
+            , "verbose", "vi", "xtrace" ]
 
     bashism t@(T_SimpleCommand id _ (cmd:rest)) =
         let name = fromMaybe "" $ getCommandName t

